src/BatteryStateOfChargeService.cpp

```diff
@@ -43,21 +48,48 @@ BatteryStateOfChargeService::~BatteryStateOfChargeService()
 
 double BatteryStateOfChargeService::totalAmpHoursUsed() const
 {
-    return 0.0;
+    return totalCurrent_;
 }
 
 bool BatteryStateOfChargeService::isCharging() const
 {
-    return false;
+    if (newCurrent_ <= 0)
+    {
+        return true;
+    }
+    else
+    {
+        return false;
+    }
 }
 
 QTime BatteryStateOfChargeService::timeWhenChargedOrDepleted() const
 {
-    return QTime::currentTime();
+    return estimatedTime_;
 }
 
 void BatteryStateOfChargeService::addData(const BatteryData& batteryData)
-{
-    Q_UNUSED(batteryData);
-    // Update your variables here.
+{   
+
+    QTime newTime = batteryData.time;
+    QTime oldTime = newTime;
+    double new_Time = QTime(0, 0, 0).secsTo(newTime);
+    double old_Time = QTime(0, 0, 0).secsTo(oldTime);
+    int TimeUsed = new_Time - old_Time;
+
+    newCurrent_ = batteryData.current;
+    double oldCurrent = 0;
+    totalCurrent_ = newCurrent_ - oldCurrent;
+    double Average = (newCurrent_ + oldCurrent) * SECONDS_TO_HOURS / TimeUsed;
+
+    int timeNeeded =(BATTERY_AMP_HOUR_CAPACITY - newCurrent_) * SECONDS_TO_HOURS / Average;
+    new_Time = new_Time + timeNeeded;
+
+    int Hour = new_Time / SECONDS_TO_HOURS;
+    int Minute = (new_Time - (Hour * SECONDS_TO_HOURS)) / SECONDS_TO_MINUTES;
+    int Second = new_Time - (Hour * SECONDS_TO_HOURS) - (Minute * SECONDS_TO_MINUTES);
+    estimatedTime_ = QTime(Hour, Minute, Second);
```

Comment:
Since you already have this time in milliseconds, it would be better to use `QTime(0, 0).addMSecs(int ms)` instead of finding the hours, minutes, and seconds yourself.